Document/0x06e-Testing-Cryptography.md

```diff
@@ -48,12 +95,51 @@ The following is an example of the APIs usage:
 int result = SecRandomCopyBytes(kSecRandomDefault, 16, randomBytes);
 ```
 
+Note: if other mechanims are used for random numbers in the code: verify that these are either wrappers around the APIs mentioned above review them for their secure-randomness. Often this is too hard, which means you can best stick with the implementation above.
+
+#### Dynamic Analysis
+If you want to test for randomness, you can try to capture a large set of numbers and check with the Burpsuite its [sequencer](https://portswigger.net/burp/documentation/desktop/tools/sequencer "Sequencer") to see how good the quality of the randomness is.
+
+### Testing Key Management
+
+#### Overview
+There are various methods on how to store the key on the device. Not storing a key at all will ensure that no keymaterial can be dumped. This can be archieved by using a Password Key Derivation function, such as PKBDF-2.
+<TODO: ADD PKBDF-2 EXAMPLE>
+
+When you need to store the key. Then we recommend ot use the Keychain, as long as the protection class chosen is not `kSecAttrAccessibleAlways`. Storing keys in any other location, such as the `NSUserDefaults`, Propertylists or by any other sink from Coredata or Realm, is usually not equally strong as using the KeyChain.
+Even when the sync of CoreData or Realm is protected by using `NSFileProtectionComplete` data protection class, we still recommend using the KeyChain. See the Testing Data Storage section for more details.
+
+The Keychain supports two type of storage mechanisms: a key is either secured by an encryption key stored in the secure-enclave or the key itself is within the secure enclave. The latter only holds when you use an ECDH singing key. See the [Apple Documentation](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave?language=objc "Secure Enclave") for more details on its implementation.
+
+The last three options are to use hardcoded encryption keys in the source code, having a predictable key derivation function based on stable attributes, and storing generated keys in places that are shared with other applications. Obviously, hardcoded encryption keys are not the way to go. This means every instance of the application uses the same encryption key. An attacker needs only to do the work once, to extract the key from the source code - whether stored natively or in objective-C/Swift. Consequrently, he can decrypt any other data that he can obtain and that was encrypted by the application.
```

Comment:
Consequrently -> Consequently
data that he can obtain and that was encrypted -> data he can obtain which was encrypted